src/libraries/Common/src/System/Reflection/Metadata/TypeNameParserHelpers.cs

```diff
@@ -0,0 +1,416 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Buffers;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Text;
+
+using static System.Char;
+using static System.Int32;
+
+#nullable enable
+
+namespace System.Reflection.Metadata
+{
+    internal static class TypeNameParserHelpers
+    {
+        internal const int MaxArrayRank = 32;
+        internal const sbyte SZArray = -1;
+        internal const sbyte Pointer = -2;
+        internal const sbyte ByRef = -3;
+        private const char EscapeCharacter = '\\';
+#if NET8_0_OR_GREATER
+        private static readonly SearchValues<char> _endOfFullTypeNameDelimitersSearchValues = SearchValues.Create("[]&*,+\\");
+#endif
+
+        /// <returns>
+        /// <para>Negative value for invalid type names.</para>
+        /// <para>Zero for valid non-generic type names.</para>
+        /// <para>Positive value for valid generic type names.</para>
+        /// </returns>
+        internal static int GetGenericArgumentCount(ReadOnlySpan<char> fullTypeName)
```

Comment:
IMO this method has more complicated control flow than necessary, including multiple return statements that can be collapsed into a smaller number of check. It would be good to simplify and say that you're ultimately only looking for four things:

1. The string ends with a sequence of ASCII digits.
2. The sequence of ASCII digits does not begin with a leading zero and fits cleanly into an int32.
3. The sequence of ASCII digits is preceded by a backtick character.
4. The backtick character isn't the first character in the string.

If you lead with this context, then the implementation reads a bit more straightforward:

```cs
static int GetGenericArgumentCount(ReadOnlySpan<char> fullTypeName)
{
    char leadingDigit = '0';
    int i;
    for (i = fullTypeName.Length - 1; i >= 0; i--)
    {
        char thisChar = fullTypeName[i];
        if (!char.IsAsciiDigit(thisChar)) { break; }
        leadingDigit = thisChar;
    }
    
    // At this point, i is the index of the last char which is not an ASCII digit,
    // or -1 if the string consists entirely of ASCII digit characters.
    
    if (leadingDigit != '0' && i >= 1 && fullTypeName[i] == '`')
    {
        return int.TryParse(fullTypeName.Slice(i + 1), NumberStyles.None, CultureInfo.InvariantCulture, out int value) ? value : -1;
    }
    
    return 0;
}
```

See also ECMA-335, Sec. I.10.7.2. In theory you should be checking for multiple backticks and rejecting those payloads, but perhaps that's an exercise best left to the caller of this method.